`ng add single-spa-angular@4` fails on fresh angular@10 project

### Demonstration
```
> npm i -g @angular/cli@10.0.3
> ng new ngcli1003
> cd .\ngcli1003\
```
```
> ng add single-spa-angular@4
Package has unmet peer dependencies. Adding the package may not succeed.
Installing packages for tooling via npm.
Installed packages for tooling via npm.
? Does your application use Angular routing? Yes
? Does your application use BrowserAnimationsModule? Yes
    Added 'single-spa' as a dependency
    Added 'single-spa-angular' as a dependency
    Added '@angular-builders/custom-webpack' as a dependency
    Generated 'main.single-spa.ts
    Generated 'single-spa-props.ts
    Generated asset-url.ts
    Generated extra-webpack.config.js
    Using @angular-devkit/custom-webpack builder.
Unexpected token / in JSON at position 0
```

### Expected Behavior
`ng add single-spa-angular@4` should work for fresh project generated via @angular/cli@10

### Actual Behavior
`ng add single-spa-angular@4` fails with the following error:
```
Unexpected token / in JSON at position 0
```

----

After some debugging, I found that it fails while parsing the tsconfig
https://github.com/single-spa/single-spa-angular/blob/cc96477d0f526a61c6603a535a0d4b07741a416c/schematics/ng-add/index.ts#L159-L167
just because all tsconfig's in @angular@10 project contains `/* comments */`.
```
> cat .\tsconfig.app.json
/* To learn more about this file see: https://angular.io/config/tsconfig. */
{
  "extends": "./tsconfig.base.json",
  "compilerOptions": {
    "outDir": "./out-tsc/app",
    "types": []
  },
  "files": [
    "src/main.ts",
    "src/polyfills.ts"
  ],
  "include": [
    "src/**/*.d.ts"
  ]
}
```
tsconfig is json5 so comments are a valid thing https://github.com/json5/json5#comments

This is another piece that would be nice to resolve in order to fully support #232

Also found that `ng add` inserts `"@angular-builders/custom-webpack": "^10"` 
https://github.com/single-spa/single-spa-angular/blob/cc96477d0f526a61c6603a535a0d4b07741a416c/schematics/ng-add/dependencies.ts#L40-L49
but there is no `@angular-builders/custom-webpack@10` right now so this part will be a problem for some time.

Tried `^9` but get

```
> spa-angular@0.0.0 build:single-spa:spa-angular C:\dev\playground\single-spa\spa-angular
> ng build spa-angular --prod


chunk {} main.js (main) 4.9 kB [entry] [rendered]
Date: 2020-07-20T23:02:28.274Z - Hash: f00cafcb1a71515b40f9 - Time: 4766ms

ERROR in Cannot read property 'flags' of undefined
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! spa-angular@0.0.0 build:single-spa:spa-angular: `ng build spa-angular --prod`
npm ERR! Exit status 1
```

Works with the latest beta `10.0.0-beta.1` tho.

--- a/.gitignore
+++ b/.gitignore
@@ -42,3 +42,6 @@ parcel/
 
 dist
 dist-server
+
+# Jetbrains ide
+.idea

--- a/integration/portal/package.json
+++ b/integration/portal/package.json
@@ -3,7 +3,7 @@
   "version": "0.0.0",
   "scripts": {
     "start": "webpack-dev-server",
-    "prebuild:prod": "rm -rf dist",
+    "prebuild:prod": "rimraf dist",
     "build:prod": "webpack -p"
   },
   "dependencies": {
@@ -15,6 +15,7 @@
     "typescript": "file:../../node_modules/typescript",
     "webpack": "^4.42.1",
     "webpack-cli": "^3.3.11",
-    "webpack-dev-server": "^3.10.3"
+    "webpack-dev-server": "^3.10.3",
+    "rimraf": "^2.6.2"
   }
 }

--- a/schematics/ng-add/index.ts
+++ b/schematics/ng-add/index.ts
@@ -29,6 +29,12 @@ import {
   getSingleSpaAngularDependency,
   getAngularBuildersCustomWebpackDependency,
 } from './dependencies';
+import {
+  findPropertyInAstObject,
+  insertPropertyInAstObjectInOrder,
+  removePropertyInAstObject,
+} from '@schematics/angular/utility/json-utils';
+import { JsonParseMode, parseJsonAst } from '@angular-devkit/core';
 
 interface CustomWebpackBuilderOptions extends BrowserBuilderOptions {
   customWebpackConfig: {
@@ -158,12 +164,29 @@ function updateProjectNewAngular(
 
 function updateTSConfig(host: Tree, clientProject: WorkspaceProject): void {
   const tsConfigFileName = clientProject.architect!.build!.options.tsConfig;
-  const tsConfig = host.read(tsConfigFileName)!.toString('utf-8');
-  const json = JSON.parse(tsConfig);
-  json.files = ['src/main.single-spa.ts'];
+  const buffer = host.read(tsConfigFileName);
+  if (!buffer) {
+    return;
+  }
+
+  const tsCfgAst = parseJsonAst(buffer.toString(), JsonParseMode.Loose);
+  if (tsCfgAst.kind !== 'object') {
+    return;
+  }
+
+  const files = findPropertyInAstObject(tsCfgAst, 'files');
+  if (files && files.kind !== 'array') {
+    return;
+  }
+
   // The "files" property will only contain path to `main.single-spa.ts` file,
   // because we remove `polyfills` from Webpack `entry` property.
-  host.overwrite(tsConfigFileName, JSON.stringify(json, null, 2));
+  const recorder = host.beginUpdate(tsConfigFileName);
+  if (files) {
+    removePropertyInAstObject(recorder, tsCfgAst, 'files');
+  }
+  insertPropertyInAstObjectInOrder(recorder, tsCfgAst, 'files', ['src/main.single-spa.ts'], 2);
+  host.commitUpdate(recorder);
 }
 
 export function addNPMScripts(options: NgAddOptions): Rule {

--- a/schematics/ng-add/tests/index.spec.ts
+++ b/schematics/ng-add/tests/index.spec.ts
@@ -3,6 +3,9 @@ import { getFileContent } from '@schematics/angular/utility/test';
 import { join, normalize } from 'path';
 
 import { Schema as NgAddOptions } from '../schema';
+import { JsonParseMode, parseJsonAst } from '@angular-devkit/core';
+import { findPropertyInAstObject } from '@schematics/angular/utility/json-utils';
+import { JsonAstObject } from '@angular-devkit/core/src/json/interface';
 
 const collectionPath = join(__dirname, '../../schematics.json');
 
@@ -22,6 +25,16 @@ const defaultApplicationOptions = {
   style: 'scss',
 };
 
+const angular10Comment = '/* Unexpected comment from Angular */';
+
+// Simulate what angular 10 is doing adding a comment to the tsconfig file
+// https://github.com/single-spa/single-spa-angular/issues/249
+function appendCommentToTsConfig(tree: UnitTestTree) {
+  let content = getFileContent(tree, '/projects/ss-angular-cli-app/tsconfig.app.json');
+  content = angular10Comment + '\n' + content;
+  tree.overwrite('/projects/ss-angular-cli-app/tsconfig.app.json', content);
+}
+
 describe('ng-add', () => {
   let defaultAppTree: UnitTestTree;
   const testRunner = new SchematicTestRunner('single-spa-angular', collectionPath);
@@ -39,6 +52,8 @@ describe('ng-add', () => {
         workspaceTree,
       )
       .toPromise();
+
+    appendCommentToTsConfig(defaultAppTree);
   });
 
   test('should run ng-add', async () => {
@@ -151,10 +166,33 @@ describe('ng-add', () => {
       .runSchematicAsync<NgAddOptions>('ng-add', { routing: true }, defaultAppTree)
       .toPromise();
 
-    const tsConfigAppJSON = JSON.parse(
-      getFileContent(tree, '/projects/ss-angular-cli-app/tsconfig.app.json'),
-    );
+    const expectedTsConfigPath = normalize('projects/ss-angular-cli-app/tsconfig.app.json');
+    const buffer = defaultAppTree.read(expectedTsConfigPath);
+    if (!buffer) {
+      throw new Error('Failed to read the tsconfig');
+    }
 
-    expect(tsConfigAppJSON.files).toEqual(['src/main.single-spa.ts']);
+    const tsCfgAst = parseJsonAst(buffer.toString(), JsonParseMode.Loose);
+    if (tsCfgAst.kind != 'object') {
+      throw new Error(`Root content of '${expectedTsConfigPath}' is not an object.`);
+    }
+
+    // We verify that we didn't erased the comment
+    if (tsCfgAst.comments === undefined) {
+      throw new Error(`No comment has been found in the '${expectedTsConfigPath}' file.`);
+    }
+    expect(tsCfgAst.comments).toHaveLength(1);
+    expect(tsCfgAst.comments[0].text).toBe(angular10Comment);
+
+    // We verify the files property
+    const files = findPropertyInAstObject(tsCfgAst, 'files');
+    if (!files) {
+      throw new Error("'files' field of tsconfig should exist.");
+    }
+    if (files.kind !== 'array') {
+      throw new Error("'files' field of tsconfig should be an array.");
+    }
+
+    expect(files.value).toStrictEqual(['src/main.single-spa.ts']);
   });
 });

